panic: interface conversion: interface {} is nil, not map[string]interface

---
name: 🐛 Bug Report
about: Error occurring (panic: interface conversion: interface {} is nil, not map[string]interface) when applying a plan in TFE

---

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
- Terraform v 1.1.5 
- provider.aws v4.14.0

### Affected Resource(s)
replaced name with EXAMPLE for privacy

- aws_msk_cluster

### Terraform Configuration Files
```hcl
terraform {
  required_providers {
    aws = {
      version = "~> 4.0"
    }
  }
  required_version = "~>1.1.5"
}


resource "aws_msk_cluster" "<EXAMPLE>-msk-cluster" {
  broker_node_group_info {
    az_distribution = "DEFAULT"
    client_subnets  = var.vpc_subnets[var.envr]
    ebs_volume_size = "1000"
    instance_type   = "kafka.m5.large"
    security_groups = var.vpc_security_groups[var.envr]
  }

  client_authentication {
    sasl {
      iam   = "true"
      scram = "true"
    }
    tls {
      certificate_authority_arns = null
    }
  }

  cluster_name = "<EXAMPLE>-msk-${var.msk_cluster_env_name[var.envr]}-001"

  encryption_info {
     encryption_at_rest_kms_key_arn = data.aws_kms_key.aws_kms_kafka.arn

    encryption_in_transit {
      client_broker = "TLS"
      in_cluster    = "true"
    }
  }

  enhanced_monitoring = var.msk_cluster_enhanced_monitoring[var.envr]
  kafka_version       = "2.6.0"

  logging_info {
    broker_logs {
      cloudwatch_logs {
        enabled   = "true"
        log_group = var.msk_cluster_log_name[var.envr]
      }

      firehose {
        enabled = "false"
      }

      s3 {
        enabled = "false"
      }
    }
  }

  number_of_broker_nodes = "3"

  open_monitoring {
    prometheus {
      jmx_exporter {
        enabled_in_broker = "true"
      }

      node_exporter {
        enabled_in_broker = "true"
      }
    }
  }

  tags = {
    "ConfigBy" = "TF: <EXAMPLE>-kafka-${var.envr}"
    "Expires"  = "2023-05-02"
    "Name"     = "<EXAMPLE>-msk-${var.msk_cluster_env_name[var.envr]}-001"
    "SME"      = "teamname"
  }
}
```
### Debug Output

### Panic Output

```
Stack trace from the terraform-provider-aws_v4.14.0_x5 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 113 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/kafka.expandClientAuthentication(0xc0022c3e00)
github.com/hashicorp/terraform-provider-aws/internal/service/kafka/cluster.go:918 +0x374
github.com/hashicorp/terraform-provider-aws/internal/service/kafka.resourceClusterUpdate({0xa4b7c20, 0xc002138a80}, 0xc002166500, {0x7f68540, 0xc000a86000})
github.com/hashicorp/terraform-provider-aws/internal/service/kafka/cluster.go:773 +0x36d7
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xa4b7c20, {0xa4b7c20, 0xc002138a80}, 0xd, {0x7f68540, 0xc000a86000})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/resource.go:736 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000ec02a0, {0xa4b7c20, 0xc002138a80}, 0xc00214add0, 0xc002166380, {0x7f68540, 0xc000a86000})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/resource.go:847 +0x9e5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00000cae0, {0xa4b7b78, 0xc001ec9640}, 0xc001dc3d60)
github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/grpc_provider.go:1021 +0xe3c
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0018f8f00, {0xa4b7c20, 0xc002138480}, 0xc000fc79d0)
github.com/hashicorp/terraform-plugin-go@v0.9.0/tfprotov5/tf5server/server.go:812 +0x56b
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x91626e0, 0xc0018f8f00}, {0xa4b7c20, 0xc002138480}, 0xc001c0b380, 0x0)
github.com/hashicorp/terraform-plugin-go@v0.9.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001fcc40, {0xa5c3290, 0xc000c121a0}, 0xc0021327e0, 0xc0018f3860, 0x101abd00, 0x0)
google.golang.org/grpc@v1.45.0/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc0001fcc40, {0xa5c3290, 0xc000c121a0}, 0xc0021327e0, 0x0)
google.golang.org/grpc@v1.45.0/server.go:1619 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
google.golang.org/grpc@v1.45.0/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc@v1.45.0/server.go:919 +0x294

Error: The terraform-provider-aws_v4.14.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Expected Behavior
Plan looks clean in terraform enterprise.
On apply in Terraform Enterprise, we should have been apply two changes; 
1.  Update in place of cluster changing the client authentication-->sasl--> iam = "true"

### Actual Behavior

panic error on apply

### Steps to Reproduce

Our apply process occurs in a Terraform Enterprise workspace. 
1. `terraform apply`

### Important Factoids
Cluster was originally configured through the UI on AWS.  
Created the terraform retroactively and imported it to state.  Now, trying to update the configuration with tf.  
Errors occured in the plan until we updated to aws.provider 4.13.0.
Plans clean now.  
We get the same apply error for both 4.13.0 and 4.14.0 

### References
This was the issue report that helped me figure out that I needed to use 4.13.0.  
https://github.com/hashicorp/terraform-provider-aws/issues/20956#issuecomment-1119109677

I can reproduce this with a test configuration very similar to the one provided:

```
╷
│ Error: Plugin did not respond
│ 
│   with aws_msk_cluster.test,
│   on main.tf line 85, in resource "aws_msk_cluster" "test":
│   85: resource "aws_msk_cluster" "test" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵

Stack trace from the terraform-provider-aws plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 95 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/kafka.expandClientAuthentication(0xc0023823c0)
	/Users/ewbankkit/src/github.com/terraform-providers/terraform-provider-aws/internal/service/kafka/cluster.go:1093 +0x374
github.com/hashicorp/terraform-provider-aws/internal/service/kafka.resourceClusterUpdate({0xb1782e0, 0xc002208e10}, 0xc00220f380, {0x8c02340, 0xc000297500})
	/Users/ewbankkit/src/github.com/terraform-providers/terraform-provider-aws/internal/service/kafka/cluster.go:889 +0x3c77
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xb1782e0, {0xb1782e0, 0xc002208e10}, 0xd, {0x8c02340, 0xc000297500})
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/resource.go:736 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000f74540, {0xb1782e0, 0xc002208e10}, 0xc001aa1ba0, 0xc00220ee00, {0x8c02340, 0xc000297500})
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/resource.go:847 +0x9e5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0001a5a40, {0xb178238, 0xc00206c980}, 0xc002009ef0)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/schema/grpc_provider.go:1021 +0xe3c
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001a39b80, {0xb1782e0, 0xc002208630}, 0xc001c11030)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.9.0/tfprotov5/tf5server/server.go:812 +0x56b
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x9e0a4c0, 0xc001a39b80}, {0xb1782e0, 0xc002208630}, 0xc00202c8a0, 0x0)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.9.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002a6700, {0xb284bb0, 0xc0009bf6c0}, 0xc002203560, 0xc001a4d5c0, 0x10ef1040, 0x0)
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc0002a6700, {0xb284bb0, 0xc0009bf6c0}, 0xc002203560, 0x0)
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:1619 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.45.0/server.go:919 +0x294

Error: The terraform-provider-aws plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/.changelog/25072.txt
+++ b/.changelog/25072.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_msk_cluster: Prevent crash on apply when `client_authentication.tls` is empty
+```

--- a/docs/contributing/data-handling-and-conversion.md
+++ b/docs/contributing/data-handling-and-conversion.md
@@ -619,7 +619,7 @@ To read:
 func expandStructure(tfMap map[string]interface{}) *service.Structure {
     // ...
 
-    if v, ok := tfMap["nested_attribute_name"].([]interface{}); ok && len(v) > 0 {
+    if v, ok := tfMap["nested_attribute_name"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
         apiObject.NestedAttributeName = expandStructure(v[0].(map[string]interface{}))
     }
 

--- a/internal/service/kafka/cluster.go
+++ b/internal/service/kafka/cluster.go
@@ -973,7 +973,7 @@ func expandBrokerNodeGroupInfo(tfMap map[string]interface{}) *kafka.BrokerNodeGr
 		apiObject.ClientSubnets = flex.ExpandStringSet(v)
 	}
 
-	if v, ok := tfMap["connectivity_info"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["connectivity_info"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.ConnectivityInfo = expandConnectivityInfo(v[0].(map[string]interface{}))
 	}
 
@@ -993,7 +993,7 @@ func expandBrokerNodeGroupInfo(tfMap map[string]interface{}) *kafka.BrokerNodeGr
 		}
 	}
 
-	if v, ok := tfMap["storage_info"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["storage_info"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.StorageInfo = expandStorageInfo(v[0].(map[string]interface{}))
 	}
 
@@ -1007,7 +1007,7 @@ func expandConnectivityInfo(tfMap map[string]interface{}) *kafka.ConnectivityInf
 
 	apiObject := &kafka.ConnectivityInfo{}
 
-	if v, ok := tfMap["public_access"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["public_access"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.PublicAccess = expandPublicAccess(v[0].(map[string]interface{}))
 	}
 
@@ -1021,7 +1021,7 @@ func expandStorageInfo(tfMap map[string]interface{}) *kafka.StorageInfo {
 
 	apiObject := &kafka.StorageInfo{}
 
-	if v, ok := tfMap["ebs_storage_info"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["ebs_storage_info"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.EbsStorageInfo = expandEBSStorageInfo(v[0].(map[string]interface{}))
 	}
 
@@ -1035,7 +1035,7 @@ func expandEBSStorageInfo(tfMap map[string]interface{}) *kafka.EBSStorageInfo {
 
 	apiObject := &kafka.EBSStorageInfo{}
 
-	if v, ok := tfMap["provisioned_throughput"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["provisioned_throughput"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.ProvisionedThroughput = expandProvisionedThroughput(v[0].(map[string]interface{}))
 	}
 
@@ -1085,11 +1085,11 @@ func expandClientAuthentication(tfMap map[string]interface{}) *kafka.ClientAuthe
 
 	apiObject := &kafka.ClientAuthentication{}
 
-	if v, ok := tfMap["sasl"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["sasl"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.Sasl = expandSasl(v[0].(map[string]interface{}))
 	}
 
-	if v, ok := tfMap["tls"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["tls"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.Tls = expandTLS(v[0].(map[string]interface{}))
 	}
 
@@ -1166,7 +1166,7 @@ func expandEncryptionInfo(tfMap map[string]interface{}) *kafka.EncryptionInfo {
 
 	apiObject := &kafka.EncryptionInfo{}
 
-	if v, ok := tfMap["encryption_in_transit"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["encryption_in_transit"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.EncryptionInTransit = expandEncryptionInTransit(v[0].(map[string]interface{}))
 	}
 
@@ -1204,7 +1204,7 @@ func expandLoggingInfo(tfMap map[string]interface{}) *kafka.LoggingInfo {
 
 	apiObject := &kafka.LoggingInfo{}
 
-	if v, ok := tfMap["broker_logs"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["broker_logs"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.BrokerLogs = expandBrokerLogs(v[0].(map[string]interface{}))
 	}
 
@@ -1218,15 +1218,15 @@ func expandBrokerLogs(tfMap map[string]interface{}) *kafka.BrokerLogs {
 
 	apiObject := &kafka.BrokerLogs{}
 
-	if v, ok := tfMap["cloudwatch_logs"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["cloudwatch_logs"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.CloudWatchLogs = expandCloudWatchLogs(v[0].(map[string]interface{}))
 	}
 
-	if v, ok := tfMap["firehose"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["firehose"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.Firehose = expandFirehose(v[0].(map[string]interface{}))
 	}
 
-	if v, ok := tfMap["s3"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["s3"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.S3 = expandS3(v[0].(map[string]interface{}))
 	}
 
@@ -1298,7 +1298,7 @@ func expandOpenMonitoringInfo(tfMap map[string]interface{}) *kafka.OpenMonitorin
 
 	apiObject := &kafka.OpenMonitoringInfo{}
 
-	if v, ok := tfMap["prometheus"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["prometheus"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.Prometheus = expandPrometheusInfo(v[0].(map[string]interface{}))
 	}
 
@@ -1312,11 +1312,11 @@ func expandPrometheusInfo(tfMap map[string]interface{}) *kafka.PrometheusInfo {
 
 	apiObject := &kafka.PrometheusInfo{}
 
-	if v, ok := tfMap["jmx_exporter"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["jmx_exporter"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.JmxExporter = expandJmxExporterInfo(v[0].(map[string]interface{}))
 	}
 
-	if v, ok := tfMap["node_exporter"].([]interface{}); ok && len(v) > 0 {
+	if v, ok := tfMap["node_exporter"].([]interface{}); ok && len(v) > 0 && v[0] != nil {
 		apiObject.NodeExporter = expandNodeExporterInfo(v[0].(map[string]interface{}))
 	}
 

